### PR TITLE
[feat] Todo list: CRUD

### LINKED ISSUE #2: [feat] Todo list: CRUD
Create base endpoints to perform CRUD operations:

- Create todo list
- Read all todo lists
- Read one todo list
- Update todo list
- Delete todo list

Example model:
Title
Description (optional)
List of linked todo items
Create date
Update date

### PR DESCRIPTION
**Pull request by dromaria**
#2

### CODE REVIEW

File: app/Http/Controllers/TodoController.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+use App\Http\Requests\StoreTodoRequest;
+use App\Http\Requests\UpdateTodoRequest;
+use App\Http\Resources\TodoResource;
+use App\Models\Todo;
+use Illuminate\Http\Resources\Json\AnonymousResourceCollection;
+
+class TodoController extends Controller
+{
+    /**
+     * Display a listing of the resource.
```
Comment (north70) [inline app/Http/Controllers/TodoController.php]: this is a useless comment. Here and in other methods

File: app/Http/Controllers/TodoController.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        $data = $request->validated();
+        $todo->update($data);
+        return new TodoResource($todo);
+    }
+
+    /**
+     * Remove the specified resource from storage.
+     */
+    public function destroy(Todo $todo)
```
Comment (north70) [inline app/Http/Controllers/TodoController.php]: return type

File: app/Http/Requests/StoreTodoRequest.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    /**
+     * Get the validation rules that apply to the request.
+     *
+     * @return array<string, ValidationRule|array|string>
+     */
+    public function rules(): array
+    {
+        return [
+            'title'=>['required', 'string'],
+            'description'=>'string',
```
Comment (north70) [inline app/Http/Requests/StoreTodoRequest.php]: use an array to define the rules

File: app/Http/Requests/StoreTodoRequest.php [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+     */
+    public function authorize(): bool
+    {
+        return true;
+    }
+
+    /**
+     * Get the validation rules that apply to the request.
+     *
+     * @return array<string, ValidationRule|array|string>
```
Comment (north70) [inline app/Http/Requests/StoreTodoRequest.php]: useless phpdoc

File: app/Http/Requests/UpdateTodoRequest.php [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+     */
+    public function authorize(): bool
+    {
+        return true;
+    }
+
+    /**
+     * Get the validation rules that apply to the request.
+     *
+     * @return array<string, ValidationRule|array|string>
```
Comment (north70) [inline app/Http/Requests/UpdateTodoRequest.php]: useless phpdoc

File: app/Http/Requests/UpdateTodoRequest.php [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+     */
+    public function rules(): array
+    {
+        return [
+            'title'=>'string',
```
Comment (north70) [inline app/Http/Requests/UpdateTodoRequest.php]: use an array to define the rules

File: app/Http/Resources/TodoResource.php [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+use Illuminate\Http\Resources\Json\JsonResource;
+
+/* @mixin Todo*/
+
+class TodoResource extends JsonResource
+{
+    /**
+     * Transform the resource into an array.
+     *
+     * @return array<string, mixed>
```
Comment (north70) [inline app/Http/Resources/TodoResource.php]: useless phpdoc and comment

File: app/Http/Resources/TodoResource.php [hunk 8]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace App\Http\Resources;
+
+use App\Models\Todo;
+use Illuminate\Http\Request;
+use Illuminate\Http\Resources\Json\JsonResource;
+
+/* @mixin Todo*/
+
```
Comment (north70) [inline app/Http/Resources/TodoResource.php]: extra line break

File: app/Models/Todo.php [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Builder;
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\SoftDeletes;
+
+/**
+ *
```
Comment (north70) [inline app/Models/Todo.php:12]: extra line breaks

File: app/Models/Todo.php [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ * @property int $id
+ * @property string $title
+ * @property string|null $description
+ * @property \Illuminate\Support\Carbon|null $created_at
```
Comment (north70) [inline app/Models/Todo.php]: reduce this through import

File: app/Models/User.php [hunk 11]
```diff
@@ -7,6 +7,33 @@
 use Illuminate\Foundation\Auth\User as Authenticatable;
 use Illuminate\Notifications\Notifiable;
 
+/**
+ * 
+ *
+ * @property int $id
+ * @property string $name
+ * @property string $email
+ * @property \Illuminate\Support\Carbon|null $email_verified_at
```
Comment (north70) [inline app/Models/User.php]: reduce through import

File: composer.json [hunk 12]
```diff
@@ -7,9 +7,11 @@
     "require": {
         "php": "^8.2",
         "laravel/framework": "^11.0",
+        "laravel/sanctum": "^4.0",
```
Comment (north70) [inline composer.json:10]: why ?

File: config/sanctum.php [hunk 13]
```diff
@@ -0,0 +1,83 @@
+<?php
```
Comment (north70) [inline config/sanctum.php:1]: why ?

File: database/migrations/2024_05_02_105031_create_personal_access_tokens_table.php [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+use Illuminate\Support\Facades\Schema;
+
+return new class extends Migration
+{
+    /**
+     * Run the migrations.
+     */
+    public function up(): void
+    {
+        Schema::create('personal_access_tokens', function (Blueprint $table) {
```
Comment (north70) [inline database/migrations/2024_05_02_105031_create_personal_access_tokens_table.php:14]: Are you sure this is what you need?

File: routes/api.php [hunk 15]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+use App\Http\Controllers\TodoController;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Route;
+
+Route::get('/user', function (Request $request) {
```
Comment (north70) [inline routes/api.php:7]: Why ?

File: app/Actions/Todo/StoreTodoAction.php [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+use App\Repositories\Interfaces\TodoRepositoryInterface;
+use Illuminate\Database\Eloquent\Model;
+
+class StoreTodoAction
+{
+    private TodoRepositoryInterface $todoRepository;
+
+    public function __construct(TodoRepositoryInterface $todoRepository)
+    {
+        $this->todoRepository = $todoRepository;
```
Comment (north70) [inline app/Actions/Todo/StoreTodoAction.php]: read the [article](https://stitcher.io/blog/constructor-promotion-in-php-8) and replace here and further

File: app/Repositories/TodoRepositories.php [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+use App\Models\Todo;
+use App\Repositories\Interfaces\TodoRepositoryInterface;
+use Illuminate\Database\Eloquent\Collection;
+use Illuminate\Database\Eloquent\Model;
+
+class TodoRepositories implements TodoRepositoryInterface
+{
+    public function index(): Collection|Todo
+    {
+        return Todo::all();
```
Comment (north70) [inline app/Repositories/TodoRepositories.php]: it's ok when you use method all, but often in a production app we use pagination with limit and offset params. Please, add them to the route

File: config/sanctum.php (hunk 13 above)
Comment (dromaria) [reply config/sanctum.php:1]: laravel does this by default

File: composer.json (hunk 12 above)
Comment (dromaria) [reply composer.json:10]: laravel does this by default

File: app/Repositories/TodoRepositories.php [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    public function store($data): Model|Todo
```
Comment (north70) [inline app/Repositories/TodoRepositories.php]: there isn't data type here. You are using the data as an array. I can provide any data and break your method. In such cases, you need to use dto-class

File: app/Repositories/TodoRepositories.php [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        return Todo::create($data);
+    }
+
+    public function update($todo, $data): Model|Todo
```
Comment (north70) [inline app/Repositories/TodoRepositories.php]: there isn't data type here. Use dto

File: database/migrations/2024_05_02_105031_create_personal_access_tokens_table.php (hunk 14 above)
Comment (dromaria) [reply database/migrations/2024_05_02_105031_create_personal_access_tokens_table.php:14]: it was automatically added when I created api.php

File: app/Repositories/TodoRepositories.php [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        $todo->update($data);
+        return $todo;
+    }
+    public function destroy($todo): Model|Todo
```
Comment (north70) [inline app/Repositories/TodoRepositories.php]: there isn't data type here

File: routes/api.php (hunk 15 above)
Comment (dromaria) [reply routes/api.php:7]: same

File: app/Actions/Todo/UpdateTodoAction.php [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+
+use App\Models\Todo;
+use App\Repositories\Interfaces\TodoRepositoryInterface;
+use Illuminate\Database\Eloquent\Model;
+
+class UpdateTodoAction
+{
+    private TodoRepositoryInterface $todoRepository;
+
+    public function __construct(TodoRepositoryInterface $todoRepository)
```
Comment (north70) [inline app/Actions/Todo/UpdateTodoAction.php]: the same situation as above

File: app/Actions/Todo/IndexTodoAction.php [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+namespace App\Actions\Todo;
+
+use App\Models\Todo;
+use App\Repositories\Interfaces\TodoRepositoryInterface;
+use Illuminate\Database\Eloquent\Collection;
+
+class IndexTodoAction
+{
+    private TodoRepositoryInterface $todoRepository;
+    public function __construct(TodoRepositoryInterface $todoRepository)
```
Comment (north70) [inline app/Actions/Todo/IndexTodoAction.php]: the same situation as down

File: app/Repositories/TodoRepositories.php [hunk 23]
```diff
@@ -2,18 +2,19 @@
...
+use Illuminate\Contracts\Pagination\LengthAwarePaginator;
 use Illuminate\Database\Eloquent\Model;
 
 class TodoRepositories implements TodoRepositoryInterface
 {
-    public function index(): Collection|Todo
+    public function index(PaginationDTO $paginationDTO): LengthAwarePaginator|Todo
     {
-        return Todo::all();
+        return Todo::paginate($paginationDTO->limit, ['*'], 'page', $paginationDTO->offset);
```
Comment (north70) [inline app/Repositories/TodoRepositories.php]: You don't have two paginations, it's an overhead. Use function offset() with simplePaginate() or just paginate()

File: app/Http/Requests/Pagination/PaginationRequest.php [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+{
+    public function authorize(): bool
+    {
+        return true;
+    }
+
+    public function rules(): array
+    {
+        return [
+            'limit'=>'int',
```
Comment (north70) [inline app/Http/Requests/Pagination/PaginationRequest.php]: Add restrictions on numbers. Now it is possible to transfer negative numbers. For offset too

File: app/Http/Requests/Pagination/PaginationRequest.php [hunk 25]
```diff
@@ -0,0 +1,31 @@
+<?php
+
+namespace App\Http\Requests\Pagination;
+
+use Illuminate\Foundation\Http\FormRequest;
+
+class PaginationRequest extends FormRequest
+{
+    public function authorize(): bool
```
Comment (north70) [inline app/Http/Requests/Pagination/PaginationRequest.php]: You don't have to write this in every class, but create a base class, specify this function in it, and extends from it everywhere

**Review by north70 — APPROVED** [review]

**Pull request merged**